Lombiq.ChartJs.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs

```diff
@@ -0,0 +1,147 @@
+    {
+        await context.GoToHistoryAsync();
+
+        context.TestChartJsChart("LineChart", "line_chart", 10);
+    }
+
+    private static void TestChartJsChart(
+        this UITestContext context,
+        string logHeader,
+        string referenceResourceName,
+        double pixelErrorThreshold)
+    {
+        // This is to avoid Chart.js animation related issues
+        var hash = context.WaitChartJsCanvasToBeReadyAndHash();
+        hash.ShouldNotBeNullOrEmpty();
+        context.Scope.AtataContext.Log.Trace($"{logHeader}: imageHash: {hash}");
+        var canvas = context.Driver.FindElementByTagName("canvas");
+        canvas.ShouldNotBeNull();
+        using var canvasImage = context.TakeScreenshotImage(canvas)
+            .ToImageSharpImage();
+        canvasImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_canvas.bmp");
+        using var referenceImage = GetResourceImageSharpImage($"Lombiq.ChartJs.Tests.UI.Assets.{referenceResourceName}.dib");
+        referenceImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_reference.bmp");
+        using var diffImage = ImageSharpCompare.CalcDiffMaskImage(
+            $"Temp/{logHeader}_canvas.bmp",
+            $"Temp/{logHeader}_reference.bmp");
+        diffImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_diff.bmp");
+        var diff = ImageSharpCompare.CalcDiff(
+            $"Temp/{logHeader}_canvas.bmp",
+            $"Temp/{logHeader}_reference.bmp");
+        context.Scope.AtataContext.Log.Trace($@"{logHeader}: diff:
+    absoluteError={diff.AbsoluteError.ToString(CultureInfo.InvariantCulture)},
+    meanError={diff.MeanError.ToString(CultureInfo.InvariantCulture)},
+    pixelErrorCount={diff.PixelErrorCount.ToString(CultureInfo.InvariantCulture)},
+    pixelErrorPercentage={diff.PixelErrorPercentage.ToString(CultureInfo.InvariantCulture)}
+");
+        diff.PixelErrorPercentage.ShouldBeLessThan(pixelErrorThreshold);
+    }
+
+    private static string ComputeSha256Hash(byte[] raw)
+    {
+        using var sha256Hash = SHA256.Create();
+
+        return string.Concat(
+            sha256Hash.ComputeHash(raw)
+                .Select(item => item.ToString("x2", CultureInfo.InvariantCulture)));
+    }
+
+    private static string ComputeElementImageHash(this UITestContext context, IWebElement element)
+    {
+        using var elementImage = context.TakeScreenshotImage(element);
+        using var elementImageStream = new MemoryStream();
+
+        elementImage.Save(elementImageStream, ImageFormat.Bmp);
+        var elementImageRaw = elementImageStream.ToArray();
+        return ComputeSha256Hash(elementImageRaw);
+    }
+
+    private static string WaitChartJsCanvasToBeReadyAndHash(
+        this UITestContext context,
+        double timeoutSec = 30,
+        double pollMillisec = 100)
+    {
+        var wait = new WebDriverWait(context.Driver, timeout: TimeSpan.FromSeconds(timeoutSec))
```

Comment:
Rather use `context.DoWithRetriesOrFail`

Reply:
I would store the result hash in log

Reply:
Could you elaborate? Why does that require using WebDriverWait?